### PR TITLE
Refactor InBandKeyManager#encrypt to operate on MemoryRecords

### PR DESCRIPTION
**Pull request by robobario**
### Type of change

- Refactoring
### Description

Why:
We are preparing to make the encrypt aware of the batches within a MemoryRecords instance and preserve the batches along with their metadata. We need to change encrypt to take MemoryRecords and return a CompletionStage<MemoryRecords> so that the batch-preservation can be handled while we encrypt all the records.

Contributes towards #855

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Review performance test results. Ensure that any degradations to performance numbers are understood and justified.
- [ ] Make sure all Sonar-Lint warnings are addressed or are justifiably ignored.
- [ ] Update documentation
- [ ] Reference relevant issue(s) and close them after merging
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

### CODE REVIEW

File: kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/InBandKeyManager.java [hunk 1]
```diff
@@ -110,18 +117,57 @@ private CompletableFuture<KeyContext> makeKeyContext(@NonNull K kekId) {
...
-        return attemptEncrypt(topicName, partition, encryptionScheme, records, receiver, 0);
+        return result.thenApply(unused -> builder.build());
+    }
+
+    @NonNull
+    private static Stream<Record> recordStream(MemoryRecords memoryRecords) {
+        return StreamSupport.stream(memoryRecords.records().spliterator(), false);
+    }
+
+    private static MemoryRecordsBuilder recordsBuilder(@NonNull ByteBufferOutputStream buffer, @NonNull MemoryRecords records) {
```
Comment (robobario) [inline kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/InBandKeyManager.java:150]: note this is just pulling the existing code into the KeyManager, the MemoryRecords produced will not have the same batching as the input. That can be done later.

File: kroxylicious-filters/kroxylicious-encryption/src/test/java/io/kroxylicious/filter/encryption/EnvelopeEncryptionFilterTest.java [hunk 2]
```diff
@@ -191,42 +184,12 @@ void shouldOnlyEncryptTopicWithKeyId() {
...
         verify(keyManager).encrypt(any(), anyInt(), any(),
-                argThat(records -> assertThat(records)
+                argThat(records -> assertThat(records.records())
                         .hasSize(1)
                         .allSatisfy(record -> assertThat(record.value()).isEqualTo(ByteBuffer.wrap(HELLO_CIPHER_WORLD)))),
                 any());
     }
 
-    @Test
-    void produceShouldMaintainClientOffsets() {
```
Comment (robobario) [inline kroxylicious-filters/kroxylicious-encryption/src/test/java/io/kroxylicious/filter/encryption/EnvelopeEncryptionFilterTest.java:201]: this is now a KeyManager responsibility, I added an assertion to check that non consecutive offsets are preserved.

File: kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/InBandKeyManager.java [hunk 3]
```diff
@@ -110,18 +117,57 @@ private CompletableFuture<KeyContext> makeKeyContext(@NonNull K kekId) {
...
-                                         @NonNull Receiver receiver) {
-        if (records.isEmpty()) {
-            return CompletableFuture.completedFuture(null);
+    public CompletionStage<MemoryRecords> encrypt(@NonNull String topicName,
+                                                  int partition,
+                                                  @NonNull EncryptionScheme<K> encryptionScheme,
+                                                  @NonNull MemoryRecords records,
+                                                  @NonNull Function<Integer, ByteBufferOutputStream> bufferAllocator) {
+        if (records.firstBatch() == null) {
+            return CompletableFuture.completedFuture(records);
```
Comment (robobario) [inline kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/InBandKeyManager.java:129]: What should we complete with here? null or the input MemoryRecords? Both work I think, but with null we would need a check in the Filter to not call set on the partitionProduceData if the result is null.

File: kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/KeyManager.java [hunk 4]
```diff
@@ -21,18 +24,18 @@
 public interface KeyManager<K> {
 
     /**
-     * Asynchronously encrypt the given {@code recordRequests} using the current DEK for the given KEK, calling the given receiver for each encrypted record
+     * Asynchronously encrypt the given {@code records} using the current DEK for the given KEK returning a transformed record batch
```
Comment (robobario) [inline kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/KeyManager.java:27]: talking about "record batch" here is incorrect, memory records contains multiple batches

File: kroxylicious-filters/kroxylicious-encryption/src/test/java/io/kroxylicious/filter/encryption/inband/InBandKeyManagerTest.java [hunk 5]
```diff
@@ -210,14 +222,14 @@ void encryptionRetry() {
 
         var value = ByteBuffer.wrap(new byte[]{ 1, 2, 3 });
         var value2 = ByteBuffer.wrap(new byte[]{ 4, 5, 6 });
-        Record record = RecordTestUtils.record(value);
-        Record record2 = RecordTestUtils.record(value2);
+        Record record = RecordTestUtils.record(value, 0L);
```
Comment (robobario) [inline kroxylicious-filters/kroxylicious-encryption/src/test/java/io/kroxylicious/filter/encryption/inband/InBandKeyManagerTest.java:225]: I found we need to start setting offsets to construct multi-record MemoryRecords from existing records, if they aren't monotonically increasing then it barfs when you try to build the MemoryRecords.

File: kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/KeyManager.java [hunk 6]
```diff
@@ -21,18 +24,18 @@
...
-    CompletionStage<Void> encrypt(@NonNull String topicName,
-                                  int partition,
-                                  @NonNull EncryptionScheme<K> encryptionScheme,
-                                  @NonNull List<? extends Record> records,
-                                  @NonNull Receiver receiver);
+    CompletionStage<MemoryRecords> encrypt(@NonNull String topicName,
+                                           int partition,
+                                           @NonNull EncryptionScheme<K> encryptionScheme,
+                                           @NonNull MemoryRecords records,
+                                           @NonNull Function<Integer, ByteBufferOutputStream> bufferAllocator);
```
Comment (robobario) [inline kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/KeyManager.java:38]: Sonar pointing out this could be IntFunction<ByteBufferOutputStream>

File: kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/InBandKeyManager.java [hunk 7]
```diff
@@ -110,18 +117,57 @@ private CompletableFuture<KeyContext> makeKeyContext(@NonNull K kekId) {
...
+                                                  @NonNull MemoryRecords records,
+                                                  @NonNull IntFunction<ByteBufferOutputStream> bufferAllocator) {
+        if (records.firstBatch() == null) {
+            return CompletableFuture.completedFuture(records);
+        }
+        MemoryRecordsBuilder builder = recordsBuilder(allocateBufferForEncode(records, bufferAllocator), records);
+        var encryptionRequests = recordStream(records).toList();
+        CompletionStage<Void> result;
+        if (encryptionRequests.isEmpty()) {
+            result = CompletableFuture.completedFuture(null);
```
Comment (robobario) [inline kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/InBandKeyManager.java:135]: This might be redundant now that we check for a null firstBatch, not 100% sure, the tests showed that an empty memory records would have a null firstbatch and it fails when you try to calculate the size for buffer allocation.

File: kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java [hunk 8]
```diff
@@ -85,6 +86,19 @@ public static Record record(ByteBuffer value,
...
+     * Return a Record with the given value, offset and headers
+     * @param value
+     * @param offset
+     * @param headers
+     * @return The record
+     */
+    public static Record record(ByteBuffer value,
+                                long offset,
+                                Header... headers) {
+        return record(RecordBatch.CURRENT_MAGIC_VALUE, offset, 0, null, value, headers);
```
Comment (tombentley) [inline kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java:99]: We should probably factor out constants for `DEFAULT_TIMESTAMP`, `DEFAULT_OFFSET` etc.

File: kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/InBandKeyManager.java (hunk 3 above)
Comment (tombentley) [reply kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/InBandKeyManager.java:129]: Let's return the `records`.

File: kroxylicious-filters/kroxylicious-encryption/src/test/java/io/kroxylicious/filter/encryption/inband/InBandKeyManagerTest.java [hunk 9]
```diff
@@ -98,6 +101,15 @@ void shouldEncryptRecordValue() {
                 .isEqualTo(value);
     }
 
+    @NonNull
+    private static CompletionStage<Void> doEncrypt(InBandKeyManager<UUID, InMemoryEdek> km, String topic, int partition, EncryptionScheme<UUID> scheme,
```
Comment (tombentley) [inline kroxylicious-filters/kroxylicious-encryption/src/test/java/io/kroxylicious/filter/encryption/inband/InBandKeyManagerTest.java:105]: Parameter per-line

File: kroxylicious-filters/kroxylicious-encryption/src/test/java/io/kroxylicious/filter/encryption/inband/InBandKeyManagerTest.java (hunk 5 above)
Comment (tombentley) [reply kroxylicious-filters/kroxylicious-encryption/src/test/java/io/kroxylicious/filter/encryption/inband/InBandKeyManagerTest.java:225]: I wonder... the RTU is specifically for tests.. perhaps there mileage in having factories which specifically return batches with non-contiguous offsets. Right now the defaults are kind-of dangerous.. they make is easy to write tests with built-in assumptions (e.g. that the offset is always zero, that offsets are contiguous, that batches are small, that the magic is always latest). What many of the tests actually need is coverage outside such defaults. So it feels like a higher-level builder would be useful.

File: kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/InBandKeyManager.java (hunk 3 above)
Comment (robobario) [reply kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/InBandKeyManager.java:129]: done, softened the javadoc to say it completes with MemoryRecords, rather than specifying that it is a new instance

File: kroxylicious-filters/kroxylicious-encryption/src/test/java/io/kroxylicious/filter/encryption/inband/InBandKeyManagerTest.java (hunk 5 above)
Comment (robobario) [reply kroxylicious-filters/kroxylicious-encryption/src/test/java/io/kroxylicious/filter/encryption/inband/InBandKeyManagerTest.java:229]: yeahp, the amount of combinations is already getting quite explosive in RTU, could imagine having a builder for Records or MemoryRecords would be a nice future PR.

Comment (tombentley) [reply kroxylicious-filters/kroxylicious-encryption/src/test/java/io/kroxylicious/filter/encryption/inband/InBandKeyManagerTest.java:229]: For consistency we should have the `offset` before the `value`.

File: kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/InBandKeyManager.java (hunk 7 above)
Comment (robobario) [reply kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/InBandKeyManager.java]: After checking with Tom there is a case where there can be a non-null batch with zero records in it, it's possible for a compacted topic to have all the records compacted out of a batch but the batch metadata needs to be preserved. We need to take care with this, for now we short-circuit respond if there are no records.

File: kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java (hunk 8 above)
Comment (robobario) [reply kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java]: introduced constants for the default record properties

File: kroxylicious-filters/kroxylicious-encryption/src/test/java/io/kroxylicious/filter/encryption/inband/InBandKeyManagerTest.java (hunk 5 above)
Comment (robobario) [reply kroxylicious-filters/kroxylicious-encryption/src/test/java/io/kroxylicious/filter/encryption/inband/InBandKeyManagerTest.java]: create #900 to capture RTU improvements

File: kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java [hunk 10]
```diff
@@ -71,7 +81,7 @@ public static String recordKeyAsString(Record record) {
      */
     public static Record record(byte[] value,
```
Comment (k-wall) [inline kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java:82]: not necessarily for now, but "record" (in method names and formal parameters) is conflicting with Java keyword.

File: kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java [hunk 11]
```diff
@@ -295,7 +308,54 @@ public static MemoryRecords memoryRecords(byte magic, long offset, long timestam
...
+        }
+    }
+
+    /**
+     * Return a MemoryRecords containing a single RecordBatch containing multiple Records.
+     * The batch will use the current magic.
+     * @param records
+     * @return The MemoryRecords
+     */
+    public static MemoryRecords memoryRecords(Collection<Record> records) {
```
Comment (k-wall) [inline kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java:323]: I would expect that records should be ordered (i.e. List rather than Collection).

File: kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java [hunk 12]
```diff
@@ -295,7 +308,54 @@ public static MemoryRecords memoryRecords(byte magic, long offset, long timestam
...
-        try (MemoryRecordsBuilder memoryRecordsBuilder = new MemoryRecordsBuilder(
+        try (MemoryRecordsBuilder memoryRecordsBuilder = defaultMemoryRecordsBuilder(magic)) {
+            memoryRecordsBuilder.appendWithOffset(offset, timestamp, key, value, headers);
+            return memoryRecordsBuilder.build();
+        }
+    }
+
+    /**
+     * Return a MemoryRecords containing a single RecordBatch containing multiple Records.
+     * The batch will use the current magic.
```
Comment (k-wall) [inline kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java:319]: you've said it will use the current magic, but that's not what the method impl does (it uses the default).

File: kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java [hunk 13]
```diff
@@ -295,7 +308,54 @@ public static MemoryRecords memoryRecords(byte magic, long offset, long timestam
     }
 
     private static MemoryRecords memoryRecordsWithoutCopy(byte magic, long offset, long timestamp, byte[] key, byte[] value, Header... headers) {
```
Comment (k-wall) [inline kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java:310]: I'm not understanding the significance of `WithoutCopy`.  Does appendWithOffset copy the (key, value, header) bytes to the destination buffer before the method returns? Specifically I see the code calling `org.apache.kafka.common.record.DefaultRecord#writeTo` so I think whether it is copied or not is irrelevant.   What am I missing?

File: kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java [hunk 14]
```diff
@@ -295,7 +308,54 @@ public static MemoryRecords memoryRecords(byte magic, long offset, long timestam
...
+            records.forEach(record -> memoryRecordsBuilder.appendWithOffset(record.offset(), record));
+            return memoryRecordsBuilder.build();
+        }
+    }
+
+    /**
+     * It is possible for a batch to have 100% of the contained records compacted away. This simulates
+     * that case.
+     */
+    public static MemoryRecords memoryRecordsWithAllRecordsRemoved() {
```
Comment (k-wall) [inline kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java:334]: why did you chose that name rather than emptyBatch?

File: kroxylicious-filter-test-support/src/test/java/io/kroxylicious/test/record/RecordTestUtilsTest.java [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+
+class RecordTestUtilsTest {
+
+    @Test
+    void testMemoryRecordsWithAllRecordsDeleted() {
+        MemoryRecords records = RecordTestUtils.memoryRecordsWithAllRecordsRemoved();
+        assertThat(records.firstBatch()).isNotNull();
+        assertThat(records.sizeInBytes()).isGreaterThan(0);
+        assertThat(records.records().iterator().hasNext()).isFalse();
```
Comment (k-wall) [inline kroxylicious-filter-test-support/src/test/java/io/kroxylicious/test/record/RecordTestUtilsTest.java:21]:
```suggestion
        assertThat(records.records()).isEmpty();
```

File: kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/KeyManager.java [hunk 16]
```diff
@@ -21,18 +24,18 @@
 public interface KeyManager<K> {
 
     /**
-     * Asynchronously encrypt the given {@code recordRequests} using the current DEK for the given KEK, calling the given receiver for each encrypted record
+     * Asynchronously encrypt the given {@code records} using the current DEK for the given KEK returning a transformed MemoryRecords
```
Comment (k-wall) [inline kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/KeyManager.java:27]: "a transformed MemoryRecords" makes it sound like MemoryRecords the original is mutated.  I don't think that's the desired contract we mean.

```suggestion
     * Asynchronously encrypt the given {@code records} using the current DEK for the given KEK returning a MemoryRecords object which will contain all records transformed according to the {@code encryptionScheme}.
```

**Review by k-wall — APPROVED** [review]
couple of nits and a couple of questions

File: kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/KeyManager.java [hunk 17]
```diff
@@ -21,18 +24,18 @@
...
-                                  @NonNull List<? extends Record> records,
-                                  @NonNull Receiver receiver);
+    CompletionStage<MemoryRecords> encrypt(@NonNull String topicName,
+                                           int partition,
+                                           @NonNull EncryptionScheme<K> encryptionScheme,
+                                           @NonNull MemoryRecords records,
+                                           @NonNull IntFunction<ByteBufferOutputStream> bufferAllocator);
 
     /**
      * Asynchronously decrypt the given {@code kafkaRecords} (if they were, in fact, encrypted), calling the given {@code receiver} with the plaintext
```
Comment (k-wall) [inline kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/KeyManager.java:41]: Should decypt operate on MemoryRecords too?

File: kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java (hunk 13 above)
Comment (robobario) [reply kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java:312]: This was from @tombentley, I think from the comment in one of the usages:
```
// No need to copy the arrays because their contents are written to a ByteBuffer and not retained
```
It's trying to describe something about the way the arrays is used. But as said I don't think it's relevant.

File: kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java (hunk 12 above)
Comment (robobario) [reply kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java:321]: Hmm, the default == current, this was a way to make that consistent over the methods. Though I take your point since the default constant is public. WDYT @tombentley

File: kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java (hunk 13 above)
Comment (robobario) [reply kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java:312]: (think I'll leave it for a separate round of RecordTestUtils cleanup/improvements)

File: kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java (hunk 12 above)
Comment (robobario) [reply kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java:321]: I've made the consts private for now.

File: kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java (hunk 10 above)
Comment (tombentley) [reply kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java:84]: It's not a keyword, it's a restricted identifier. Personally I remain to be convinced that these sonar warnings have any value, or that fixing them is not a waste of effort. Can anyone think of a realistic usage where someone reading the code it going to get confused by the fact that `record` means one thing in a type declaration and something else when used as a variable name?

File: kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java (hunk 13 above)
Comment (tombentley) [reply kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java:312]: I found some nasty corner cases in testing tests which made me much more defensive about this than I was initially. 

* If the caller passed `byte[]` they could still mutate the contents
* If the caller passed `ByteBuffer`, they could still mutate the contents or the position etc. 

In order to minimise the time wasted investigating tests which are doing weird things I think we want the `Records` returned from this method to be immutable.
